### PR TITLE
HPCC-20978 Ensure ,DISTRIBUTE creates a distributed dataset when possible

### PR DESCRIPTION
**Pull request by ghalliday**
Also adds support for ,DISTRIBUTED for some other forms of inline dataset

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review

**Comment by ghalliday** [comment]
Documentation jira is HPCC-21054

File: testing/regress/ecl/complexhoist3.ecl [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+############################################################################## */
+
+//HOIST(dataset({unsigned i}) ds) := NOFOLD(SORT(NOFOLD(ds), i));
+HOIST( ds) := MACRO
+//NOFOLD(SORT(NOFOLD(ds), i))
+//NOFOLD(ds)
+ds
+ENDMACRO;
+
+mkRow(unsigned value) := TRANSFORM({ unsigned i }, SKIP(value = 1000000);   SELF.i := value);
```
Comment (shamser) [inline testing/regress/ecl/complexhoist3.ecl]: @ghalliday Where is this transform used? Don't see it used in this file.

File: testing/regress/ecl/complexhoist2.ecl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+//HOIST(dataset({unsigned i}) ds) := NOFOLD(SORT(NOFOLD(ds), i));
+HOIST( ds) := MACRO
+//NOFOLD(SORT(NOFOLD(ds), i))
+//NOFOLD(ds)
+ds
+ENDMACRO;
+
+mkRow(unsigned value) := TRANSFORM({ unsigned i }, SKIP(value = 1000000);   SELF.i := value);
+
+dsOuter  := HOIST(DATASET(3, mkRow(COUNTER)));
```
Comment (shamser) [inline testing/regress/ecl/complexhoist2.ecl]: @ghalliday Doesn't "distributed" have to be added to the end to have this distributed?

**Review by shamser — COMMENTED** [review]
@ghalliday I'd like some clarification on the regression code.  Is there another Jira where the actual work on distributed datasets is implemented?

File: testing/regress/ecl/complexhoist2.ecl (hunk 2 above)
Comment (ghalliday) [reply testing/regress/ecl/complexhoist2.ecl]: Yes, this is a set of tests I added to try and provoke a problem in the selector processing.  Not all the tests are directly relevant.

File: testing/regress/ecl/complexhoist3.ecl (hunk 1 above)
Comment (ghalliday) [reply testing/regress/ecl/complexhoist3.ecl]: Agreed, it could be cleaned up

**Comment by ghalliday** [comment]
The code was already implemented for the syntax DATASET(<n>, transform), but not for the syntax DATASET([...],...)
They end up generating the same activities, which is why there are so few changes.

**Comment by ghalliday** [comment]
@shamser forgot to reassign

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed.

File: ecl/hql/hqlgram.y [hunk 3]
```diff
@@ -9159,15 +9159,16 @@ simpleDataSet
                             $$.setExpr(createDataset(no_inlinetable, createValue(no_transformlist, makeNullType(), values), createComma(LINK(record), $7.getExpr())));
                             $$.setPosition($1);
                         }
-    | DATASET '(' thorFilenameOrList ',' beginCounterScope recordDef endCounterScope ')'
+    | DATASET '(' thorFilenameOrList ',' beginCounterScope dsRecordDef endCounterScope optDatasetFlags dsEnd
```
Comment (richardkchapman) [inline ecl/hql/hqlgram.y]: Should the documentation be updated to reflect this new option?

**Pull request merged**